README.md
# Curitiba - Dados Abertos
## Arquivo limpo do dadosbertos - Portal curitiba.pr.gov.br

[![Build Status](https://travis-ci.org/CodeForCuritiba/ds-curitiba-dados-abertos.svg?branch=master)](https://travis-ci.org/CodeForCuritiba/ds-curitiba-dados-abertos)

Esse repositório visa abranger todos os dados do portal da transparência de maneira limpa e pronta para ser utilizada por programas de BI. [](https://www.curitiba.pr.gov.br/dadosabertos/busca/)
O mesmo gera entregáveis que podem ser acessados facilmente na aba "releases" deste mesmo repositório. Este é um projeto piloto do code for Curitiba que visa utilizar técnicas de CI/CD.

## Como acessar os dados
Esse projeto foi construído em cima do `docker-composer`. Este permite a criação de um ambiente pronto para funcionamento em todas as plataformas: Windows, Linux e Mac OS;
O instalador para Windows pode ser encontrado em [](https://docs.docker.com/docker-for-windows/).

## Instalação e primeiro acesso
Para instalar esse repositório, faça download dos arquivos ou clone este repositório GIT e, dentro do diretório dele, execute os seguintes comandos.

    docker-compose up -d

Aponte o navegador para `http://localhost:8888` e pronto! Você terá um sistema com o jupyter notebook instalado. [](https://jupyter-notebook.readthedocs.io/en/stable/notebook.html)
## Contribua!
Dúvidas? Está afim de contribuir? Por favor crie um issue nesse projeto!

## Mantenedor
Esse projeto está sendo inicialmente mantido por:

* Jon Trigueiro <joepreludian at gmail dot com>


CHANGELOG.md
#3.1.0
- Adicionado mais módulos na interface; 
- Criado arquivo de exemplo para execução do projeto;
- Ativado algumas bibliotecas do nbextension;
- Bugfixes; melhorado utilização do Qgrid

#3.0.0
- Modificado a plataforma para trabalhar com o pacote python curitiba-dados-abertos

#2.1.1
- Bugfixes

#2.1.0
- Atualizado projeto. Adicionado Pipeline de entrega contínua;

#2.0.0
- Nova versão utilizando biblioteca de datascience da code

#1.1.3
- Small fixes
- Update readme file

#1.1.2
- Update travis in order to build only PRS or master branch

#1.1.1
- Fix .travis.yml

# 1.1.0
- Add ability to handle docker files
- Update Makefile

# 1.0.0
- Initial version


test_app.py
import os

import pytest
from pathlib import Path

from curitiba156 import \
    get_downloaded_base156, get_latest_csv_url_from_156, download_file


@pytest.fixture
def supply_download_data():
    return {
        'url': 'https://www.google.com.br/intl/pt-BR/add_url.html',
        'temp_name': 'temp_name',
        'temp_dir': 'source_data'
    }


def test_get_url_156():
    url = get_latest_csv_url_from_156()

    assert 'Base_de_Dados.csv' in url
    # assert 'ISO-8859-1' in charset


def test_download_simple_file(supply_download_data):

    filename, encoding = download_file(url=supply_download_data['url'])

    assert filename == 'add_url.html'
    assert os.path.isfile(filename) is True

    os.unlink('add_url.html')  # Cleaning up file


def test_download_file_with_override(supply_download_data):

    filename, encoding = download_file(
        url=supply_download_data['url'],
        name_override=supply_download_data['temp_name'])

    assert filename == supply_download_data['temp_name']
    assert os.path.isfile(filename) is True

    os.unlink(filename)


def test_download_file_with_folder(supply_download_data):

    filename, encoding = download_file(
        url=supply_download_data['url'],
        folder=supply_download_data['temp_dir'])

    assert filename == f'{supply_download_data["temp_dir"]}/add_url.html'
    assert os.path.isfile(filename) is True

    os.unlink(filename)


def test_download_file_with_folder_and_name_override(supply_download_data):

    filename, encoding = download_file(
        url=supply_download_data['url'],
        folder=supply_download_data['temp_dir'],
        name_override=supply_download_data['temp_name'])

    assert filename == f'{supply_download_data["temp_dir"]}/' \
                       f'{supply_download_data["temp_name"]}'
    assert os.path.isfile(filename) is True

    os.unlink(filename)


def test_download_no_overwrite(supply_download_data):

    filename, encoding = download_file(
        url=supply_download_data['url'],
        folder=supply_download_data['temp_dir'],
        name_override=supply_download_data['temp_name'])

    assert filename == f'{supply_download_data["temp_dir"]}/' \
                       f'{supply_download_data["temp_name"]}'

    assert os.path.isfile(filename) is True

    Path(filename).touch()
    file_stats_touched = os.stat(filename)
    assert type(file_stats_touched.st_mtime) is float

    # Downloading again and comparises if the file was overwritten
    filename, encoding = download_file(
        url=supply_download_data['url'],
        folder=supply_download_data['temp_dir'],
        name_override=supply_download_data['temp_name'],
        force_overwrite=False)

    file_stats_new = os.stat(filename)
    assert file_stats_touched.st_mtime == file_stats_new.st_mtime

    os.unlink(filename)


def test_download_with_overwrite(supply_download_data):

    filename, encoding = download_file(
        url=supply_download_data['url'],
        folder=supply_download_data['temp_dir'],
        name_override=supply_download_data['temp_name'])

    assert filename == f'{supply_download_data["temp_dir"]}/' \
                       f'{supply_download_data["temp_name"]}'
    assert os.path.isfile(filename) is True

    Path(filename).touch()
    file_stats_touched = os.stat(filename)
    assert type(file_stats_touched.st_mtime) is float

    # Downloading again and comparises if the file was overwritten
    filename, encoding = download_file(
        url=supply_download_data['url'],
        folder=supply_download_data['temp_dir'],
        name_override=supply_download_data['temp_name'],
        force_overwrite=True)

    file_stats_new = os.stat(filename)
    assert file_stats_touched.st_mtime != file_stats_new.st_mtime

    os.unlink(filename)


def test_get_base156():
    filename, encoding = get_downloaded_base156()

    assert filename == 'source_data/base156.csv'
    assert os.path.isfile(filename)


main.py
import os
import codecs

import requests

import pandas as pd
from sqlalchemy import create_engine

from xml.dom import minidom


def download_file(url, folder=None, name_override=None, force_overwrite=False):
    local_filename = url.split('/')[-1] if not name_override else name_override
    local_filename = local_filename if not folder \
        else f'{folder}/{local_filename}'

    print('Downloading File...')

    file_encoding = None
    with requests.get(url, stream=True) as r:
        r.raise_for_status()
        file_encoding = r.encoding

        if os.path.isfile(local_filename):
            print('File Already exists!')

            if not force_overwrite:
                print('No download needed.')
                return local_filename, file_encoding
            else:
                print('Forcing download anyway...')

        with open(local_filename, 'wb') as file_handler:
            for chunk in r.iter_content(chunk_size=8192):
                if chunk:
                    file_handler.write(chunk)

    print('Download Sucessful')
    return local_filename, file_encoding


def get_latest_csv_url_from_156():
    print('Fetching latest base of 156...')

    base_url = 'http://dadosabertos.c3sl.ufpr.br/curitiba/156'

    xml_index_get_request = requests.get(f'{base_url}/156.xml')

    xml_index = None
    if xml_index_get_request.status_code == 200:
        xml_index = minidom.parseString(xml_index_get_request.text)
    else:
        raise SystemError('Unable to download database from Curitiba - 156')

    dados = xml_index.getElementsByTagName('dado')
    csv_file_url = None
    for dado in dados:
        name = dado.attributes['arquivo'].value
        if '.csv' in name:
            csv_file_url = name
            break

    latest_csv_url = f'{base_url}/{csv_file_url}'

    print(f'Latest csv url: {latest_csv_url}')

    return latest_csv_url


def get_base156():

    csv_url = get_latest_csv_url_from_156()
    csv_filename, csv_filename_encoding = \
        download_file(csv_url, folder='source_data',
                      name_override='base156.csv',
                      force_overwrite=False)

    print(f'File {csv_filename}; Encoding: {csv_filename_encoding}')

    return csv_filename, csv_filename_encoding


def main():
    csv_file, csv_encoding = get_base156()

    print('Loading data from CSV...')
    field_names = ['SOLICITACAO', 'TIPO', 'ORGAO', 'DATA',
                   'HORARIO', 'ASSUNTO', 'SUBDIVISAO', 'DESCRICAO',
                   'LOGRADOURO_ASS', 'BAIRRO_ASS', 'REGIONAL_ASS',
                   'MEIO_RESPOSTA', 'OBSERVACAO', 'SEXO', 'BAIRRO_CIDADAO',
                   'REGIONAL_CIDADAO', 'DATA_NASC', 'TIPO_CIDADAO',
                   'ORGAO_RESP', 'RESPOSTA_FINAL', 'RESPOSTA_FINAL_DETALHE']

    data = pd.read_csv(csv_file, sep=';', encoding=csv_encoding,
                       error_bad_lines=False, skiprows=[0, 1],
                       names=field_names)

    print('Exporting read Dataset to SQLITE...')
    sqlite_db_engine = create_engine(
        'sqlite:///clean_data/base156.sqlite', echo=False)

    data.to_sql('base156', con=sqlite_db_engine, if_exists='replace')

    data_sexo = data[['SEXO']].groupby(
        ['SEXO']).size().reset_index(name='counts')

    data_sexo.to_sql('solicitacoes_por_sexo', con=sqlite_db_engine,
                     if_exists='replace')

    data_orgao = data[['ORGAO']].groupby(
        ['ORGAO']).size().reset_index(name='counts')

    data_orgao.to_sql('solicitacoes_por_orgao', con=sqlite_db_engine,
                      if_exists='replace')


if __name__ == '__main__':
    main()


start.sh
#!/bin/bash

figlet "CODE FOR CURITIBA"
echo "Iniciando painel dados abertos";

# Instalando bibliotecas
figlet "Pacotes Python..."
pip install --upgrade pip;
pip install notebook jupyter_contrib_nbextensions requests pandarallel jupyterthemes curitiba-dados-abertos;
pip install jupyter_nbextensions_configurator;

figlet "Pacotes Jupyter...";

# Ativando extensões extras do jupyter
jupyter contrib nbextension install --sys-prefix;
jupyter nbextensions_configurator enable --sys-prefix
jupyter nbextension enable --py --sys-prefix qgrid;
jupyter nbextension enable --py --sys-prefix widgetsnbextension;
jupyter nbextension enable --sys-prefix hinterland/hinterland;
jupyter nbextension enable --sys-prefix scroll_down;
jupyter nbextension enable --sys-prefix gist_it/main;
jupyter nbextension enable --sys-prefix execute_time/ExecuteTime;
jupyter nbextension enable --sys-prefix "tree-filter/index";


figlet "Iniciando Jupyter";
echo "Carregando serviço na porta 8888"
# Customizando Jupyter Notebook - https://github.com/dunovank/jupyter-themes
jt -fs 95 -altp -tfs 11 -nfs 115 -cellw 94% -T

# Iniciando Jupyter Notebook
jupyter notebook --ip=0.0.0.0 --port=8888 --allow-root --NotebookApp.token='' --NotebookApp.password='';


curitiba156.py
import os

import requests

import pandas as pd

from xml.dom import minidom


def download_file(url, folder=None, name_override=None, force_overwrite=False):
    """ Downloads a file from a website

    params:
        - url (mandatory)
        - folder - Destination folder;
        - name_override - If provided, override the name of file,
                          if not pick latest part of url
        - force_overwrite - (bool) - Default False
    returns:
        - local_filename - The path of downloaded file
        - file_encoding
    """
    local_filename = url.split('/')[-1] if not name_override else name_override
    local_filename = local_filename if not folder \
        else f'{folder}/{local_filename}'

    print('Downloading File...')

    file_encoding = None
    with requests.get(url, stream=True) as r:
        r.raise_for_status()
        file_encoding = r.encoding

        if os.path.isfile(local_filename):
            print('File Already exists!')

            if not force_overwrite:
                print('No download needed.')
                return local_filename, file_encoding
            else:
                print('Forcing download anyway...')

        with open(local_filename, 'wb') as file_handler:
            for chunk in r.iter_content(chunk_size=8192):
                if chunk:
                    file_handler.write(chunk)

    print('Download Sucessful')
    return local_filename, file_encoding


def get_latest_csv_url_from_156():
    """This fuction connects to website of dadosabertos, fetches and parses
    this file, collecting latest CSV.

    returns:
        - latest_csv_url - The full url of the latest CSV
    """
    print('Fetching latest base of 156...')

    base_url = 'http://dadosabertos.c3sl.ufpr.br/curitiba/156'

    xml_index_get_request = requests.get(f'{base_url}/156.xml')

    xml_index = None
    if xml_index_get_request.status_code == 200:
        xml_index = minidom.parseString(xml_index_get_request.text)
    else:
        raise SystemError('Unable to download database from Curitiba - 156')

    dados = xml_index.getElementsByTagName('dado')
    csv_file_url = None
    for dado in dados:
        name = dado.attributes['arquivo'].value
        if '.csv' in name:
            csv_file_url = name
            break

    latest_csv_url = f'{base_url}/{csv_file_url}'

    print(f'Latest csv url: {latest_csv_url}')

    return latest_csv_url


def get_downloaded_base156():
    """ Downloads the base156.

    returns:
        - csv_filename - The relative file location, on your filesystem
        - csv_filename_encoding
    """

    csv_url = get_latest_csv_url_from_156()
    csv_filename, csv_filename_encoding = \
        download_file(csv_url, folder='source_data',
                      name_override='base156.csv',
                      force_overwrite=False)

    print(f'File {csv_filename}; Encoding: {csv_filename_encoding}')

    return csv_filename, csv_filename_encoding


def main():
    csv_file, csv_encoding = get_downloaded_base156()


if __name__ == '__main__':
    main()


app.py
import os

import requests

import pandas as pd
from sqlalchemy import create_engine

from xml.dom import minidom


def download_file(url, folder=None, name_override=None, force_overwrite=False):
    """ Downloads a file from a website

    params:
        - url (mandatory)
        - folder - Destination folder;
        - name_override - If provided, override the name of file,
                          if not pick latest part of url
        - force_overwrite - (bool) - Default False
    returns:
        - local_filename - The path of downloaded file
        - file_encoding
    """
    local_filename = url.split('/')[-1] if not name_override else name_override
    local_filename = local_filename if not folder \
        else f'{folder}/{local_filename}'

    print('Downloading File...')

    file_encoding = None
    with requests.get(url, stream=True) as r:
        r.raise_for_status()
        file_encoding = r.encoding

        if os.path.isfile(local_filename):
            print('File Already exists!')

            if not force_overwrite:
                print('No download needed.')
                return local_filename, file_encoding
            else:
                print('Forcing download anyway...')

        with open(local_filename, 'wb') as file_handler:
            for chunk in r.iter_content(chunk_size=8192):
                if chunk:
                    file_handler.write(chunk)

    print('Download Sucessful')
    return local_filename, file_encoding


def get_latest_csv_url_from_156():
    """This fuction connects to website of dadosabertos, fetches and parses
    this file, collecting latest CSV.

    returns:
        - latest_csv_url - The full url of the latest CSV
    """
    print('Fetching latest base of 156...')

    base_url = 'http://dadosabertos.c3sl.ufpr.br/curitiba/156'

    xml_index_get_request = requests.get(f'{base_url}/156.xml')

    xml_index = None
    if xml_index_get_request.status_code == 200:
        xml_index = minidom.parseString(xml_index_get_request.text)
    else:
        raise SystemError('Unable to download database from Curitiba - 156')

    dados = xml_index.getElementsByTagName('dado')
    csv_file_url = None
    for dado in dados:
        name = dado.attributes['arquivo'].value
        if '.csv' in name:
            csv_file_url = name
            break

    latest_csv_url = f'{base_url}/{csv_file_url}'

    print(f'Latest csv url: {latest_csv_url}')

    return latest_csv_url


def get_downloaded_base156():
    """ Downloads the base156.

    returns:
        - csv_filename - The relative file location, on your filesystem
        - csv_filename_encoding
    """

    csv_url = get_latest_csv_url_from_156()
    csv_filename, csv_filename_encoding = \
        download_file(csv_url, folder='source_data',
                      name_override='base156.csv',
                      force_overwrite=False)

    print(f'File {csv_filename}; Encoding: {csv_filename_encoding}')

    return csv_filename, csv_filename_encoding


def main():
    csv_file, csv_encoding = get_downloaded_base156()

    print('Loading data from CSV...')
    field_names = ['SOLICITACAO', 'TIPO', 'ORGAO', 'DATA',
                   'HORARIO', 'ASSUNTO', 'SUBDIVISAO', 'DESCRICAO',
                   'LOGRADOURO_ASS', 'BAIRRO_ASS', 'REGIONAL_ASS',
                   'MEIO_RESPOSTA', 'OBSERVACAO', 'SEXO', 'BAIRRO_CIDADAO',
                   'REGIONAL_CIDADAO', 'DATA_NASC', 'TIPO_CIDADAO',
                   'ORGAO_RESP', 'RESPOSTA_FINAL', 'RESPOSTA_FINAL_DETALHE']

    data = pd.read_csv(csv_file, sep=';', encoding=csv_encoding,
                       error_bad_lines=False, skiprows=[0, 1],
                       names=field_names)

    print('Exporting read Dataset to SQLITE...')
    sqlite_db_engine = create_engine(
        'sqlite:///clean_data/base156.sqlite', echo=False)

    data.to_sql('base156', con=sqlite_db_engine, if_exists='replace')

    data_sexo = data[['SEXO']].groupby(
        ['SEXO']).size().reset_index(name='counts')

    data_sexo.to_sql('solicitacoes_por_sexo', con=sqlite_db_engine,
                     if_exists='replace')

    data_orgao = data[['ORGAO']].groupby(
        ['ORGAO']).size().reset_index(name='counts')

    data_orgao.to_sql('solicitacoes_por_orgao', con=sqlite_db_engine,
                      if_exists='replace')


if __name__ == '__main__':
    main()
